File: array/1.js
// Tc = O(n)
// Sc = O(n)
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function(nums, target) {
    let map = new Map()
    for (let i = 0;i<nums.length; i++){
        let complement = target - nums[i]
        if(map.has(complement)){
            return [map.get(complement),i]
        }
        map.set(nums[i],i)
    }
    return []
};

var twoSum = function(nums, target){
    let list = {}
    for(let i = 0; i<nums.length; i++){
        let diff = target - nums[i]
        if(list.hasOwnProperty(diff)){
            return [list[diff], i]
        }
        list[nums[i]] = i
    }
}

File: array/1.py
class Solution:
    def twoSum(self, nums: list[int], target: int) -> list[int]:
        pair_idx = {}
        for i, num in enumerate(nums):
            if target - num in pair_idx:
                return [i, pair_idx[target-num]]
            pair_idx[num] = i

class Solution:
    def twoSum(self, nums: list[int], target: int) -> list[int]:
        mapping = {}
        for i in range(len(nums)):
            diff = target - nums[i]
            if diff in mapping:
                return [mapping[diff], i]
            else:
                mapping[nums[i]] = i